lib/src/realm_class.dart

```diff
@@ -113,12 +115,56 @@ class Realm implements Finalizable {
   /// Opens a `Realm` using a [Configuration] object.
   Realm(Configuration config) : this._(config);
 
-  Realm._(this.config, [RealmHandle? handle, this._isInMigration = false]) : _handle = handle ?? _openRealm(config) {
+  Realm._(this.config, [RealmHandle? handle, this._isInMigration = false]) : _handle = handle ?? _openRealmSync(config) {
     _populateMetadata();
     isFrozen = realmCore.isFrozen(this);
   }
 
-  static RealmHandle _openRealm(Configuration config) {
+  /// A method for asynchronously opening a [Realm].
+  ///
+  /// When the configuration is [FlexibleSyncConfiguration], the realm will be downloaded and fully
+  /// synchronized with the server prior to the completion of the returned [Future].
+  /// This method could be called also for opening a local [Realm] with [LocalConfiguration].
+  ///
+  /// * `config`- a configuration object that describes the realm.
+  /// * `cancellationToken` - an optional [CancellationToken] used to cancel the operation.
+  /// * `onProgressCallback` - a callback for receiving download progress notifications for synced [Realm]s.
+  ///
+  /// Returns `Future<Realm>` that completes with the [Realm] once the remote [Realm] is fully synchronized or with a [CancelledException] if operation is canceled.
+  /// When the configuration is [LocalConfiguration] this completes right after the local [Realm] is opened or if the operation is canceled in advance.
+  /// Since opening a local Realm is a synchronous operation, there is no benefit of using Realm.open over the constructor.
+  static Future<Realm> open(Configuration config, {CancellationToken? cancellationToken, ProgressCallback? onProgressCallback}) async {
+    if (cancellationToken != null && cancellationToken.isCancelled) {
+      throw cancellationToken.exception;
+    }
+    final realm = Realm(config);
+    StreamSubscription<SyncProgress>? subscription;
+    try {
+      if (config is FlexibleSyncConfiguration) {
+        final session = realm.syncSession;
+        if (onProgressCallback != null) {
+          subscription = session
+              .getProgressStream(
+                ProgressDirection.download,
+                ProgressMode.forCurrentlyOutstandingWork,
+              )
+              .listen(
+                (syncProgress) => onProgressCallback.call(syncProgress),
+                cancelOnError: true,
+              );
+        }
+        await session.waitForDownload(cancellationToken);
+      }
+    } catch (_) {
+      realm.close();
+      rethrow;
+    } finally {
+      await subscription?.cancel();
```

Comment:
I would like this to happen before `realm.close()`, but it doesn't seem to be a problem with current core.

Reply:
Ok I change it.